design/nab_status_update.md

```diff
@@ -0,0 +1,82 @@
+# Developer Workflow: NonAdminBackup Phase and Conditions within NonAdminBackup Status
+
+## Overview
+
+This document outlines the design around updating NonAdminBackup objects Phase and Conditions within Status.
+
+### Phase
+
+A NonAdminBackup's `status` field has a `phase` field, which is updated by NAC controller.
+
+The `phase` is a simple one high-level summary of the lifecycle of an NonAdminBackup.
+
+It is always a one well defined value, that is intended to be a comprehensive state of a NonAdminBackup.
+
+Those are are the possible values for phase:
+
+| **Value** | **Description**                 |
+|-----------|--------------------------------|
+| New | *NonAdminBackup resource was accepted by the OpenShift cluster, but it has not yet been processed by the NonAdminController* |
+| BackingOff | Velero *Backup* object was not created due to NonAdminBackup error (configuration or similar) |
+| Created | Velero *Backup* was created. The Phase will not have additional informations about the |
+
+### Conditions
+
+The `conditions` is also a part of the NonAdminBackup's `status` field. One NAB object may have multiple conditions. It is more granular knowledge of the NonAdminBackup object and represents the array of the conditions through which the NonAdminBackup has or has not passed. Each `NonAdminCondition` has one of the following `type`:
+
+| **Condition** | **Description**                 |
+|-----------|--------------------------------|
+| BackupAccepted | The Backup object was accepted by the reconcile loop, but the Velero Backup may have not yet been created |
+| BackupQueued | The Velero Backup was created succesfully. At this stage errors may still occur either from the Velero not accepting backup or during backup procedure. |
+
+The `condition` data is also accomapied with the following:
+
+| **Field name** | **Description**                 |
+|-----------|--------------------------------|
+| type | The `Type` of the condition |
+| status | represents the state of individual condition. The resulting `phase` value should report `Success` only when all of the conditions are met and the backup succeded. One of `True`, `False` or `Unknown`. |
+| lastProbeTime | Timestamp of when the NonAdminBackup condition was last probed. |
+| lastTransitionTime | Timestamp for when the NonAdminBackup last transitioned from one status to another. |
+| reason | Machine-readable, UpperCamelCase text indicating the reason for the condition's last transition. |
+| message | Human-readable message indicating details about the last status transition. |
+
+### BackupStatus
+
```

Comment:
Can you please add a sample of how the NAB status in yaml form would look like ? Also the overall golang struct to give a rough idea ?

Reply:
Ok here is entire (failing) one, anything else from it?


```yaml
apiVersion: nac.oadp.openshift.io/v1alpha1
kind: NonAdminBackup
metadata:
  creationTimestamp: '2024-04-15T20:27:35Z'
  generation: 2
  managedFields:
    - apiVersion: nac.oadp.openshift.io/v1alpha1
      fieldsType: FieldsV1
      fieldsV1:
        'f:spec':
          .: {}
          'f:backupSpec': {}
      manager: Mozilla
      operation: Update
      time: '2024-04-15T20:27:35Z'
    - apiVersion: nac.oadp.openshift.io/v1alpha1
      fieldsType: FieldsV1
      fieldsV1:
        'f:spec':
          'f:backupSpec':
            'f:itemOperationTimeout': {}
            'f:metadata': {}
            'f:volumeSnapshotLocations': {}
            'f:snapshotMoveData': {}
            'f:defaultVolumesToFsBackup': {}
            'f:ttl': {}
            'f:csiSnapshotTimeout': {}
            'f:storageLocation': {}
            'f:hooks': {}
      manager: main
      operation: Update
      time: '2024-04-15T20:27:45Z'
    - apiVersion: nac.oadp.openshift.io/v1alpha1
      fieldsType: FieldsV1
      fieldsV1:
        'f:status':
          .: {}
          'f:backupStatus':
            .: {}
            'f:expiration': {}
            'f:failureReason': {}
            'f:formatVersion': {}
            'f:phase': {}
            'f:startTimestamp': {}
            'f:version': {}
          'f:conditions': {}
          'f:oadpVeleroBackup': {}
      manager: main
      operation: Update
      subresource: status
      time: '2024-04-16T08:12:12Z'
  name: testnonadminbackup
  namespace: nacproject
  resourceVersion: '61393755'
  uid: 12712cf9-2161-4e58-931d-21528afdff89
spec:
  backupSpec:
    volumeSnapshotLocations:
      - velero-sample-1
    defaultVolumesToFsBackup: false
    csiSnapshotTimeout: 10m0s
    ttl: 720h0m0s
    itemOperationTimeout: 4h0m0s
    metadata: {}
    storageLocation: velero-sample-1
    hooks: {}
    snapshotMoveData: false
status:
  backupStatus:
    expiration: '2024-05-16T08:12:11Z'
    failureReason: >-
      unable to get credentials: unable to get key for secret: Secret
      "cloud-credentials" not found
    formatVersion: 1.1.0
    phase: Failed
    startTimestamp: '2024-04-16T08:12:11Z'
    version: 1
  conditions:
    - lastTransitionTime: '2024-04-15T20:27:35Z'
      message: backup accepted
      reason: backup_accepted
      status: 'True'
      type: BackupAccepted
    - lastTransitionTime: '2024-04-15T20:27:45Z'
      message: Created Velero Backup object
      reason: backup_scheduled
      status: 'True'
      type: BackupQueued
  oadpVeleroBackup: nab-nacproject-83fc04a2fd253d
```

Reply:
This helps, so I would say we need to restructure a bit here, just remove the `backupStatus` named struct, all the feilds of `backupStatus` can be directly under `NAB.status`. This would mean:
Current traversal is: `NonAdminBackup.status.BackupStatus.phase`
Suggested/Modified traversal will be : `NonAdminBackup.status.phase`
The `backupStatus` under status seems redundant.

Reply:
I am not in favor of putting the backupStatus directly under NAB.status. It's because from implementation it's easier to compare structures that are same and NAB.status have already additional fields.

Secondly it's clear what comes from Velero Status and what is addition to that status such as `veleroBackup` or `veleroBackupNamespace` or even `conditions`. After your proposal we would have:


```yaml
status:
  expiration: '2024-05-16T08:12:11Z'
  failureReason: >-
    unable to get credentials: unable to get key for secret: Secret
    "cloud-credentials" not found
  formatVersion: 1.1.0
  phase: Failed
  startTimestamp: '2024-04-16T08:12:11Z'
  version: 1
  conditions:
    - lastTransitionTime: '2024-04-15T20:27:35Z'
      message: backup accepted
      reason: backup_accepted
      status: 'True'
      type: BackupAccepted
    - lastTransitionTime: '2024-04-15T20:27:45Z'
      message: Created Velero Backup object
      reason: backup_scheduled
      status: 'True'
      type: BackupQueued
  veleroBackup: nab-nacproject-83fc04a2fd253d
  veleroBackupNamespace: openshift-adp

```